command/sync.go

```diff
@@ -0,0 +1,501 @@
+			} else {
+				dstObject = nil
+				iDst--
+			}
+		}
+
+		switch {
+		case srcObject == nil && dstObject == nil:
+			// do nothing
+		case srcObject == nil:
+			dstOnly = append(dstOnly, dstObject.URL)
+		case dstObject == nil:
+			srcOnly = append(srcOnly, srcObject.URL)
+		}
+	}
+	return
+}
+
+// GetSourceAndDestinationObjects returns source and destination objects from given urls.
+func (s Sync) GetSourceAndDestinationObjects(ctx context.Context, srcurl, dsturl *url.URL) ([]*storage.Object, []*storage.Object, error) {
+	sourceClient, err := storage.NewClient(ctx, srcurl, s.storageOpts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	destClient, err := storage.NewClient(ctx, dsturl, s.storageOpts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var sourceObjects []*storage.Object
+	var destObjects []*storage.Object
+	var wg sync.WaitGroup
+
+	// get source objects.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		srcObjectChannel := sourceClient.List(ctx, srcurl, false)
+		for srcObject := range srcObjectChannel {
+			if s.shouldSkipObject(srcObject, true) {
+				continue
+			}
+			sourceObjects = append(sourceObjects, srcObject)
+		}
+	}()
+
+	// add * to end of destination string, to get all objects recursively.
+	var destinationURLPath string
+	if strings.HasSuffix(s.dst, "/") {
+		destinationURLPath = s.dst + "*"
+	} else {
+		destinationURLPath = s.dst + "/*"
+	}
+
+	destObjectsURL, err := url.New(destinationURLPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// get destination objects.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		destObjectsChannel := destClient.List(ctx, destObjectsURL, false)
+		for destObject := range destObjectsChannel {
+			if s.shouldSkipObject(destObject, false) {
+				continue
+			}
+			destObjects = append(destObjects, destObject)
+		}
+	}()
+
+	// wait until source and destination objects are fetched.
+	wg.Wait()
+	return sourceObjects, destObjects, nil
+}
+
+// PlanRun prepares the commands and passes it to reader for run command.
+func (s Sync) PlanRun(ctx context.Context, onlySource, onlyDest []*url.URL, common []*CommonObject,
+	dsturl *url.URL, transferManager transfer.Manager, strategy strategy.Strategy, w *io.PipeWriter, errPlan *error) {
+	// only source objects.
+	for _, srcurl := range onlySource {
+		var curDestURL *url.URL
+		switch {
+		case !srcurl.IsRemote() && dsturl.IsRemote(): // local->remote
+			curDestURL = transferManager.PrepareRemoteDestination(srcurl, dsturl)
+		case srcurl.IsRemote() && !dsturl.IsRemote(): // remote->local
+			url, err := transferManager.PrepareLocalDestination(ctx, srcurl, dsturl)
+			curDestURL = url
+			if err != nil {
+				printError(s.fullCommand, s.op, err)
+				*errPlan = multierror.Append(*errPlan, err)
+				continue
+			}
+		case srcurl.IsRemote() && dsturl.IsRemote(): // remote->remote
+			curDestURL = transferManager.PrepareRemoteDestination(srcurl, dsturl)
+		default:
+			panic("unexpected src-dst pair")
```

Comment:
I think we should only calculate destination urls in planning phase, `transfer.Manager.PrepareLocalDestination` tries to create local directories hence introduces the error handling. `cp`  would already create files/directories if we pass correct urls to it.

To calculate destination urls, seems like we only need these three bits;
https://github.com/peak/s5cmd/blob/7ba574f1838887a975e7f603c8fdc93b74cfdcb2/transfer/destination.go#L24-L27 https://github.com/peak/s5cmd/blob/7ba574f1838887a975e7f603c8fdc93b74cfdcb2/transfer/destination.go#L17-L19
https://github.com/peak/s5cmd/blob/7ba574f1838887a975e7f603c8fdc93b74cfdcb2/transfer/destination.go#L56-L58



This would also simplify the switch case as we could only check whether destination url is remote or not;
```go
for _, srcurl := range onlySource {
        // ....
    switch {
    case dsturl.IsRemote():
        // calculate remote url
    case !dsturl.IsRemote():
        // calculate local url
    default:
        // ....
    }
}
```

This way `Sync.PlanRun`'s only responsibility would be to create commands for `Run`. One might even add a `dry-run` option by consuming the pipe instead of passing it to `Run` without touching creating any directories.

Reply:
Changed as you suggested. Can you please check it from [here](https://github.com/peak/s5cmd/blob/2e00d11b005011d42c3003bdc48c940266fbe25e/command/sync.go#L358-L378)